Source/Fuse.Elements/Resources/Exif/Exif.uno

```diff
@@ -0,0 +1,242 @@
+			return GetOrientationImpl(_bytes.Length, ReadByte);
+		}
+
+		[Foreign(Language.ObjC)]
+		int GetOrientationImpl(int length, Func<int,byte> readByte)
+		@{
+
+			CGDataProviderDirectCallbacks callbacks = {
+				.getBytePointer = 0,
+				.getBytesAtPosition = getBytesAtPosition
+			};
+			UnoCallbackBox box = { readByte };
+
+			CGDataProviderRef dataProvider = CGDataProviderCreateDirect(&box, length, &callbacks);
+			CGImageSourceRef source = CGImageSourceCreateWithDataProvider(dataProvider, NULL);
+			NSArray *metadataArray = nil;
+
+			if (dataProvider) {
+				if (source) {
+					CGImageMetadataRef metadata = CGImageSourceCopyMetadataAtIndex(source, 0, NULL);
+					if (metadata) {
+						metadataArray = CFBridgingRelease(CGImageMetadataCopyTags(metadata));
+						CFRelease(metadata);
+					}
+					CFRelease(source);
+				}
+				CGDataProviderRelease(dataProvider);
+			}
+
+			int rotation = 0;
+			NSString* tagTarget = @"Orientation";
+			for (id tag in metadataArray) {
+				CFStringRef tagName = CGImageMetadataTagCopyName((CGImageMetadataTagRef)tag);
+				if ([tagTarget isEqualToString: (__bridge NSString*)tagName]) {
+					CFTypeRef rot = CGImageMetadataTagCopyValue((CGImageMetadataTagRef)tag);
+					rotation = [((__bridge NSNumber*)rot) intValue];
+				}
+			}
+
+			return rotation;
+
+		@}
+	}
+
+	[Require("Gradle.Dependency.Compile", "com.drewnoakes:metadata-extractor:2.10.1")]
+	[ForeignInclude(Language.Java,
+		"java.io.ByteArrayInputStream",
+		"java.io.InputStream",
+		"java.io.IOException",
+		"com.drew.imaging.ImageMetadataReader",
+		"com.drew.imaging.ImageProcessingException",
+		"com.drew.metadata.Metadata",
+		"com.drew.metadata.MetadataException",
+		"com.drew.metadata.exif.ExifIFD0Directory")]
+	//Uses https://drewnoakes.com/code/exif/
+	extern(Android) class ExifAndroidImpl : IExifImpl
+	{
+		byte[] _buffer;
+		public ExifAndroidImpl(byte[] buffer)
+		{
+			_buffer = buffer;
+		}
+
+		public int GetOrientation()
+		{
+			var stream = CreateStream(ReadByte);
+			return GetOrientationImpl(stream);
+		}
+
+		int _position = 0;
+		int ReadByte()
+		{
+			if (_position >= _buffer.Length)
+			{
+				return -1;
+			}
+			return (int)_buffer[_position++];
+		}
+
+		[Foreign(Language.Java)]
+		Java.Object CreateStream(Func<int> readByte)
+		@{
+			return new java.io.InputStream() {
+				public int read() {
+					return (int)readByte.run();
+				}
+			};
+		@}
+
+		[Foreign(Language.Java)]
+		int GetOrientationImpl(Java.Object stream)
+		@{
+			InputStream s = (InputStream)stream;
+			try {
+				Metadata metadata = ImageMetadataReader.readMetadata(s);
+				ExifIFD0Directory directory = metadata.getFirstDirectoryOfType(ExifIFD0Directory.class);
+				if (directory == null) {
+					return -1;
+				}
+				try {
+					int orientation = directory.getInt(ExifIFD0Directory.TAG_ORIENTATION);
+					return orientation;
+				} catch (MetadataException e) {
+				}
+			} catch (ImageProcessingException e) {
+			} catch (IOException e) {
+			}
```

Comment:
I'm unable to parse the indentation here...

But even so, do we really want to ignore all these exceptions?! That sounds dubious to me...

Reply:
I would say that we try to prevent them from happening in the first place.

I'm guessing you're getting compile-errors from Java because of missing exception-handlers here? If that's the whole deal, then just wrap the whole thing in `try { .... } catch (Exception e) { e.printStackTrace(); }`.

If you actually want to convey some information, you might want to pass an onFailed-Action in so you can report things back in a clean way.

Reply:
If you use `Directory.getInteger` rather than `Directory.getInt`, you won't have to have the catch. 

So, I guess you could write it like this to make it (IMO) a bit clearer?
```java
InputStream s = (InputStream)stream;
try {
	Metadata metadata = ImageMetadataReader.readMetadata(s);

	ExifIFD0Directory directory = metadata.getFirstDirectoryOfType(ExifIFD0Directory.class);
	if (directory != null) {
		Integer orientation = directory.getInteger(ExifIFD0Directory.TAG_ORIENTATION);
		if (orientation != null)
			return (int)orientation;
	}
} catch (Exception e) {
	e.printStackTrace();
}
return -1;
```